File: kotlin-node/karakum/src/jsMain/resources/annotations/annotateUnusedTypealiasParameter.js
import ts from "typescript"

export default (node) => {
    if (
        ts.isTypeParameterDeclaration(node)
        && node.parent
        && ts.isTypeAliasDeclaration(node.parent)
        && (
            node.parent.name.text === "WithImplicitCoercion"

            || node.parent.name.text === "PipelineCallback"
            || node.parent.name.text === "PipelineDestination"
            || node.parent.name.text === "PipelinePromise"
            || node.parent.name.text === "PipelineSource"
            || node.parent.name.text === "PipelineSourceFunction"
            || node.parent.name.text === "PipelineTransform"
            || node.parent.name.text === "PipelineTransformSource"

            || node.parent.name.text === "CustomPromisify"
        )
    ) {
        return `@Suppress("UNUSED_TYPEALIAS_PARAMETER")`
    }

    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/annotations/annotateUnusedTypealiasParameter.kt
package node.karakum.annotations

import io.github.sgrishchenko.karakum.extension.AnnotationContext
import io.github.sgrishchenko.karakum.util.getParentOrNull
import node.karakum.util.nullable
import typescript.Node
import typescript.isTypeAliasDeclaration
import typescript.isTypeParameterDeclaration

fun annotateUnusedTypealiasParameter(node: Node, context: AnnotationContext) = nullable {
    ensure(isTypeParameterDeclaration(node))

    val typeAlias = ensureNotNull(node.getParentOrNull())
    ensure(isTypeAliasDeclaration(typeAlias))
    ensure(
        typeAlias.name.text == "WithImplicitCoercion"

                || typeAlias.name.text == "PipelineCallback"
                || typeAlias.name.text == "PipelineDestination"
                || typeAlias.name.text == "PipelinePromise"
                || typeAlias.name.text == "PipelineSource"
                || typeAlias.name.text == "PipelineSourceFunction"
                || typeAlias.name.text == "PipelineTransform"
                || typeAlias.name.text == "PipelineTransformSource"

                || typeAlias.name.text == "CustomPromisify"
    )

    "@Suppress(\"UNUSED_TYPEALIAS_PARAMETER\")"
}
